torch/fx/passes/_tensorify_python_scalars.py

```diff
@@ -244,13 +247,42 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
                         )
 
                     node.replace_all_uses_with(replacement_proxy.node)
-
                     graph.erase_node(node)
 
-    # DCE symbols (which are guaranteed to be pure) only
-    for proxy in reversed(expr_to_sym_proxy.values()):
-        if len(proxy.node.users) == 0 and proxy.node.op != "placeholder":
-            graph.erase_node(proxy.node)
+    # Now do one more pass that specializes all symfloats we didn't manage
+    # to tensorify away.
+    for node in reversed(graph.nodes):
+        if node.op == "output" or node.op == "placeholder":
+            continue
+
+        with graph.inserting_before(node):
+            if len(node.users) == 0 and not node.is_impure():
+                graph.erase_node(node)
+                continue
+
+            if isinstance(
+                (val := node.meta.get("val")),
+                (torch.SymFloat, torch.SymInt, torch.SymBool),
+            ):
+                if len(val.node.expr.free_symbols) > 0 and all(
```

Comment:
what happens when there are no free symbols :0 is it already constant then?

Reply:
It feels like it's safe to run the code below anyway

Reply:
Yeah as discussed offline there's an underlying bug in bool specialization. I'll fix the bug and remove this condition. In particular this test `python test/dynamo/test_dynamic_shapes.py DynamicShapesMiscTests.test_runtime_assert_replacement_dynamic_shapes` 

without this condition results in the voided asserts without proper guarding

graph: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/bobren/ab43b2c1-6932-4de3-98d5-8773931aaa3a/custom/0_0_0/inductor_post_grad_graph_4.txt
guards: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/bobren/ab43b2c1-6932-4de3-98d5-8773931aaa3a/custom/0_0_0/dynamo_cpp_guards_str_7.txt